File: src/main/java/graph/dataAccess/ArtistDataAccess.java
package graph.dataAccess;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

import graph.models.Artist;

public class ArtistDataAccess extends DataAccess<Artist> {
  @Override
  Class<Artist> getEntityType() {
    return Artist.class;
  }

  @Override
  public Optional<Artist> getByDiscogsId(Long discogsId) { return super.getByDiscogsId(discogsId); }

  @Override
  public Iterator<Artist> getByDiscogsId(Set<Long> discogsIds) { return super.getByDiscogsId(discogsIds); }

  /**
   * Establish the HasAlias and IsAliasOf relationships between an artist and
   * an alias.
   * @param alias
   * @param artist
   */
  public void createBidirectionalAliasEdges(Artist alias, Artist artist) {
    this.setAliasOf(alias, artist);
    this.addAlias(artist, alias);
  }
  /**
   * Establish directed HasAlias edge from artist to alias.
   * The edge is added to any existing alias edges/
   * @param artist.aliases
   * @param artist
   * @param alias
   * @return
   */
  private void addAlias(Artist artist, Artist alias) {
    Set<Artist> aliases = artist.getAliases();
    aliases.add(alias);
    artist.setAliases(aliases);
  }

  /**
   * Establish the IsAliasOf edge from alias to aliasOf
   * @param alias
   * @param aliasOf
   */
  private void setAliasOf(Artist alias, Artist aliasOf) {
    alias.setAliasOf(aliasOf);
  }

  /**
   * Establish the HasMember and IsMemberOf relationships between a group member and
   * a group.
   * @param member
   * @param group
   */
  public void createBidirectionalMemberEdges(Artist member, Artist group) {
    this.setMemberOf(member, group);
    this.addMember(group, member);
  }

  /**
   * Establish the HasMember edge from group to member
   * @param group.members
   * @param group
   * @param member
   */
  private void addMember(Artist group, Artist member) {
    Set<Artist> members = group.getMembers();
    members.add(member);
    group.setMembers(members);
  }

  /**
   * Establish the IsMemberOf edge from member to group
   * @param member
   * @param group
   */
  private void setMemberOf(Artist member, Artist group) {
    member.setMemberOf(group);
  }

  public ArtistDataAccess(GetSession getSession) {
    super.getSession = getSession;
  }
}

File: src/main/scala/graph/dataAccess/ArtistDataAccess.scala
package graph.dataAccess

import scala.concurrent.Future
import neotypes.Driver
import neotypes.generic.auto._
import neotypes.implicits.syntax.all._
import graph.models.nodes.Artist
import graph.CypherUtils.{joinCypher, propertyCriteria}
import graph.queries.ArtistQueries.{
  getArtistIdentifier,
  matchArtistsQueryFragment,
  matchArtistsQueryFragment
}

class ArtistDataAccess(driver: Driver[Future]) {

  /**
    * Get a single artist, returning a map of id to Artist,
    * or an empty map if not found.
    */
  def getByDiscogsId(id: Long): Future[Map[Long, Artist]] = {
    getByDiscogsId(Set(id))
  }

  /**
    * Get multiple artists, returning a map of id to Artist
    * for the records that are found
    */
  def getByDiscogsId(ids: Set[Long]): Future[Map[Long, Artist]] = {
    val artistsCql = joinCypher(ids.map { id =>
      c"$id"
    }.toSeq)
    val query = c"MATCH (a: Artist) WHERE [$artistsCql] RETURN a.discogsId, a"
      .query[(Long, Artist)]

    query.map(driver)
  }

  /**
    * Create a single artist, without any relationships
    */
  def create(artist: Artist): Future[Unit] = {
    create(Set(artist))
  }

  /**
    * Create multiple artists, without relationships
    * e.g.
    * CREATE
    *    (foo: Artist { discogsId: $p1, name: $p2 ...}),
    *    (bar: Artist { discogsId: $p3 ...}), ...
    */
  def create(artists: Set[Artist]): Future[Unit] = {
    val artistCypherFragments = artists
      .map { artist =>
        val id = artist.discogsId
        val identifier = s"a$id"
        c"" + s"($identifier: Artist " + c"{ $artist })"
      }

    val queryBuilder = joinCypher(artistCypherFragments.toSeq)

    val query = c"CREATE $queryBuilder".query[Unit]
    query.execute(driver)
  }

  /**
    * One or both of two updates are possible: a change to the Artist itself,
    * or a change to its relationships, i.e. aliases and members.
    *  If a relationship is being added, we assume the related node already exists.
    * @param artist The node being updated
    * @param aliases Artist nodes that are aliases of this node
    * @param members Artist nodes that are members of this node
    */
  def update(
      artist: Artist,
      aliases: Set[Artist],
      members: Set[Artist]
  ): Future[Unit] = {
    createAliasRelationship(artist, aliases)
  }

  private def createAliasRelationship(artist: Artist, aliases: Set[Artist]) = {
    val artistMatchQueryBuilder = matchArtistsQueryFragment(artist.discogsId)

    // (a100: Artist {discogsId: 100})
    val aliasMatchQueryBuilder = matchArtistsQueryFragment(
      aliases.map(_.discogsId)
    )

    // i.e. (nicoleMoudaber)-[r:HasAlias]->(hairLady)
    val hasAliasQueryBuilder = joinCypher(aliases.map { alias =>
      c"" + s"(${getArtistIdentifier(artist.discogsId)})-[r${artist.discogsId}${alias.discogsId}:HasAlias]->(a${alias.discogsId})"
    }.toSeq)

    // i.e. (hairLady)-[r:IsAlias]->(nicoleMoudaber)
    val isAliasQueryBuilder = joinCypher(aliases.map { alias =>
      c"" + s"(${getArtistIdentifier(alias.discogsId)}-[r${alias.discogsId}${artist.discogsId}:IsAlias]->(a${artist.discogsId})"
    }.toSeq)

    val queryBuilder =
      c"MATCH $artistMatchQueryBuilder, $aliasMatchQueryBuilder MERGE $hasAliasQueryBuilder MERGE $isAliasQueryBuilder"

    val query = queryBuilder.query[Unit]
    println(query)
    query.execute(driver)
  }
}
